File: hackathons/workshops/payment-plans/ts/dynamic-pricing.ts
/**
 * Nevermined Lab: Payment Plans — Dynamic Pricing Server
 *
 * Charge different amounts per request based on complexity.
 * The credits parameter accepts a function instead of a fixed number.
 */

import express from "express";
import { Payments, EnvironmentName } from "@nevermined-io/payments";
import { paymentMiddleware } from "@nevermined-io/payments/express";

const payments = Payments.getInstance({
  nvmApiKey: process.env.NVM_API_KEY!,
  environment: (process.env.NVM_ENVIRONMENT || "sandbox") as EnvironmentName,
});

const PLAN_ID = process.env.NVM_PLAN_ID!;

const app = express();
app.use(express.json());

// Dynamic credits: charge based on request complexity
app.use(
  paymentMiddleware(payments, {
    "POST /generate": {
      planId: PLAN_ID,
      // Function receives (req, res) — decide cost at settlement time
      credits: (req: any) => {
        const prompt = req.body?.prompt || "";
        // 1 credit per 500 chars, minimum 1
        return Math.max(1, Math.ceil(prompt.length / 500));
      },
    },
  }) as express.RequestHandler,
);

app.post("/generate", async (req, res) => {
  const { prompt } = req.body;
  const creditsCharged = Math.max(1, Math.ceil((prompt || "").length / 500));
  res.json({
    result: `Generated response for: ${prompt}`,
    creditsCharged,
  });
});

app.listen(3000, () => {
  console.log("Dynamic pricing server on http://localhost:3000");
});

File: hackathons/workshops/payment-plans/python/dynamic_pricing.py
"""
Nevermined Lab: Payment Plans — Dynamic Pricing Server

Charge different amounts per request based on complexity.
The credits parameter accepts a function instead of a fixed number.
"""

import os
import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel
from payments_py import Payments, PaymentOptions
from payments_py.x402.fastapi import PaymentMiddleware
from dotenv import load_dotenv
import urllib3

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
load_dotenv()

payments = Payments.get_instance(
    PaymentOptions(
        nvm_api_key=os.getenv("NVM_API_KEY", ""),
        environment=os.getenv("NVM_ENVIRONMENT", "sandbox"),
    )
)

PLAN_ID = os.getenv("NVM_PLAN_ID", "")

app = FastAPI()


class GenerateRequest(BaseModel):
    prompt: str


# Dynamic credits: charge based on request complexity
app.add_middleware(
    PaymentMiddleware,
    payments=payments,
    routes={
        "POST /generate": {
            "plan_id": PLAN_ID,
            # Function receives the request — decide cost at settlement time
            "credits": lambda req: max(1, len(req.body.get("prompt", "")) // 500),
        }
    },
)


@app.post("/generate")
async def generate(body: GenerateRequest):
    credits_charged = max(1, len(body.prompt) // 500)
    return {
        "result": f"Generated response for: {body.prompt}",
        "credits_charged": credits_charged,
    }


if __name__ == "__main__":
    print("Dynamic pricing server on http://localhost:3000")
    uvicorn.run(app, host="0.0.0.0", port=3000)
